src/bindings/python/flux/job/output.py

```diff
@@ -0,0 +1,957 @@
+        if self.component:
+            prefix += f"{self.component}: "
+        if include_file_and_line and self.file:
+            prefix += f"{self.file}:{self.line}: "
+        return prefix + self.message
+
+
+class OutputHeaderEvent(EventLogEvent):
+    """
+    Object representing an RFC 24 header event
+    Attributes:
+        timestamp (float): timestamp for this event
+        name (str): Name of this event: 'header'
+        version (int): RFC 24 version
+        stdout_count (int): expected number of stdout streams
+        stderr_count (int): expected number of stderr streams
+        stdout_encoding (str): default encoding for stdout
+        stderr_encoding (str): default encoding for stderr
+        event (dict): original event as dict
+    """
+
+    def __init__(self, entry):
+        super().__init__(entry)
+        if self.name != "header":
+            raise ValueError(f"event {self.name} is not a header event")
+        self.version = self.context["version"]
+        self.stderr_count = self.context["count"]["stderr"]
+        self.stdout_count = self.context["count"]["stdout"]
+        self.stdout_encoding = self.context["encoding"]["stdout"]
+        self.stderr_encoding = self.context["encoding"]["stderr"]
+
+
+class RedirectEvent(EventLogEvent):
+    """
+    Object representing an RFC 24 redirect event
+    Attributes:
+        timestamp (float): timestamp for this event
+        name (str): Name of this event: 'redirect'
+        stream (str): the name of the stream being redirected
+        rank (Taskset): the set of ranks being redirected
+        path (str): redirect path
+        dict (dict): original event as dict
+    """
+
+    def __init__(self, entry):
+        super().__init__(entry)
+        if self.name != "redirect":
+            raise ValueError("event {self.name} is not a redirect event")
+        self.stream = self.context["stream"]
+        self.rank = Taskset(self.context["rank"])
+        self.path = self.context["path"]
+
+    def render(self):
+        return f"{self.rank}: {self.stream} redirected to {self.path}"
+
+
+class JobExceptionEvent(EventLogEvent):
+    """
+    Object representing a JobException event
+    Attributes:
+        timestamp (float): timestamp for this event
+        name (str): name of this event: 'exception'
+        severity (int): exception severity
+        exc_type (str): exception type
+        note (str): exception note
+    """
+
+    def __init__(self, entry):
+        super().__init__(entry)
+        if self.name != "exception":
+            raise ValueError("event {self.name} is not an exception event")
+        self.severity = self.context["severity"]
+        self.exc_type = self.context["type"]
+        self.note = self.context["note"]
+
+    def render(self):
+        result = f"job.exception: type={self.exc_type} severity={self.severity}"
+        if self.note:
+            result += f" {self.note}"
+        return result
```

Comment:
i was debating in my head, given note is a requirement now, do we want to output empty string note if none set??  I'm going back and forth about it ... not sure.

Reply:
While I think an empty note makes coding around job exception events cleaner, as a message to the user an empty note isn't necessary and may even be a little confusing. It isn't too important, but I'd prefer to leave it off at this point to make the output nicer.